### PR TITLE
Range of Motion Active Task

### PR DESCRIPTION
**Pull request by darrenlevy**
Special thanks to [Dr. Raj Karia](http://www.med.nyu.edu/biosketch/kariar01) and [John Guydo](http://johnguydo.tumblr.com/).

Touch Anywhere Active Task
Range of Motion Active Tasks:
- Knee
- Shoulder

The Touch Anywhere Active Task was included to allow the user to get their device in the proper position before beginning the Range of Motion Step.

Range of Motion's result contains two properties: flexed and extended. These correspond to the measurements a doctor may read when using a goniometer.

Resolves Issue https://github.com/ResearchKit/ResearchKit/issues/781

### CODE REVIEW

**Comment by darrenlevy** [comment]
Hi @YuanZhu-apple, I have rebased with master.
This PR is now compatible with ResearchKit 1.4 and Xcode 8.0.

File: ResearchKit.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -1581,6 +1606,40 @@
...
+			children = (
+				959A2C0A1D68C8F100841B04 /* Shoulder */,
+				959A2BFA1D68B98700841B04 /* ORKRangeOfMotionStep.h */,
+				959A2BFB1D68B98700841B04 /* ORKRangeOfMotionStep.m */,
+				959A2BFE1D68B9DA00841B04 /* ORKRangeOfMotionStepViewController.h */,
+				959A2BFF1D68B9DA00841B04 /* ORKRangeOfMotionStepViewController.m */,
+			);
+			name = "Range of Motion";
+			sourceTree = "<group>";
+		};
```
Comment (umerkhan-apple) [inline ResearchKit.xcodeproj/project.pbxproj]: Move the folder to the bottom of the ActiveTasks group.

File: ResearchKit/ActiveTasks/ORKRangeOfMotionStep.h [hunk 2]
```diff
@@ -0,0 +1,28 @@
+//
+//  ORKRangeOfMotionStep.h
+//  ResearchKit
+//
+//  Created by Darren Levy on 8/20/16.
+//  Copyright © 2016 researchkit.org. All rights reserved.
+//
```
Comment (umerkhan-apple) [inline ResearchKit/ActiveTasks/ORKRangeOfMotionStep.h]: Update headers to match RK legal headers. Refer to any of the other RK header files and copy/paste the legal header.

Replace Copyright Apple with your name. Look at other contributions from the past for reference on the headers.

File: ResearchKit/ActiveTasks/ORKRangeOfMotionStep.h [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+@import Foundation;
+#import <ResearchKit/ORKDefines.h>
+#import <ResearchKit/ORKActiveStep.h>
+#import <ResearchKit/ORKOrderedTask.h>
```
Comment (umerkhan-apple) [inline ResearchKit/ActiveTasks/ORKRangeOfMotionStep.h]: Review the coding guidelines.
One thing here is to have 2 new lines before imports, and 2 new lines after imports.

File: ResearchKit/ActiveTasks/ORKRangeOfMotionStep.m [hunk 4]
```diff
@@ -0,0 +1,77 @@
+//
+//  ORKRangeOfMotionStep.m
+//  ResearchKit
+//
+//  Created by Darren Levy on 8/20/16.
+//  Copyright © 2016 researchkit.org. All rights reserved.
+//
```
Comment (umerkhan-apple) [inline ResearchKit/ActiveTasks/ORKRangeOfMotionStep.m]: Update header.

File: ResearchKit/ActiveTasks/ORKRangeOfMotionStep.h [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+NS_ASSUME_NONNULL_BEGIN
+
+/**
+ The `ORKRangeOfMotionStep` class represents a step that takes a range of motion measurement.
+ */
+ORK_CLASS_AVAILABLE
+@interface ORKRangeOfMotionStep : ORKActiveStep
+
+@property (nonatomic, assign) ORKPredefinedTaskLimbOption limbOption;
```
Comment (umerkhan-apple) [inline ResearchKit/ActiveTasks/ORKRangeOfMotionStep.h]: Add some documentation for this property.

File: ResearchKit/ActiveTasks/ORKRangeOfMotionStepViewController.h [hunk 6]
```diff
@@ -0,0 +1,26 @@
+//
+//  ORKRangeOfMotionStepViewController.h
+//  ResearchKit
+//
+//  Created by Darren Levy on 8/20/16.
+//  Copyright © 2016 researchkit.org. All rights reserved.
+//
```
Comment (umerkhan-apple) [inline ResearchKit/ActiveTasks/ORKRangeOfMotionStepViewController.h]: legal header

File: ResearchKit/ActiveTasks/ORKRangeOfMotionStepViewController.h [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+#import <ResearchKit/ResearchKit.h>
+#import <CoreMotion/CMDeviceMotion.h>
```
Comment (umerkhan-apple) [inline ResearchKit/ActiveTasks/ORKRangeOfMotionStepViewController.h]: new lines

File: ResearchKit/ActiveTasks/ORKRangeOfMotionStepViewController.m [hunk 8]
```diff
@@ -0,0 +1,185 @@
+//
+//  ORKRangeOfMotionStepViewController.m
+//  ResearchKit
+//
+//  Created by Darren Levy on 8/20/16.
+//  Copyright © 2016 researchkit.org. All rights reserved.
+//
```
Comment (umerkhan-apple) [inline ResearchKit/ActiveTasks/ORKRangeOfMotionStepViewController.m]: legal header

File: ResearchKit/ActiveTasks/ORKRangeOfMotionStepViewController.m [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+#import "ORKRangeOfMotionStepViewController.h"
+#import "ORKCustomStepView_Internal.h"
+#import "ORKHelpers_Internal.h"
+#import "ORKActiveStepViewController_Internal.h"
+#import "ORKVerticalContainerView_Internal.h"
+#import "ORKDeviceMotionRecorder.h"
+#import "ORKActiveStepView.h"
+#import "ORKProgressView.h"
+#import "ORKSkin.h"
```
Comment (umerkhan-apple) [inline ResearchKit/ActiveTasks/ORKRangeOfMotionStepViewController.m]: new lines

File: ResearchKit/ActiveTasks/ORKRangeOfMotionStepViewController.m [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+    _topConstraint.constant = CaptionBaselineToProgressTop - CaptionBaselineToStepViewTop;
+}
+
+- (void)updateConstraints {
+    [self updateConstraintConstantsForWindow:self.window];
+    [super updateConstraints];
+}
+
+@end
+
```
Comment (umerkhan-apple) [inline ResearchKit/ActiveTasks/ORKRangeOfMotionStepViewController.m]: 2 new lines between implementations.

File: ResearchKit/ActiveTasks/ORKShoulderRangeOfMotionStep.h [hunk 11]
```diff
@@ -0,0 +1,17 @@
+//
+//  ORKShoulderRangeOfMotionStep.h
+//  ResearchKit
+//
+//  Created by Darren Levy on 8/20/16.
+//  Copyright © 2016 researchkit.org. All rights reserved.
+//
```
Comment (umerkhan-apple) [inline ResearchKit/ActiveTasks/ORKShoulderRangeOfMotionStep.h]: legal header

File: ResearchKit/ActiveTasks/ORKShoulderRangeOfMotionStep.h [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+
+#import "ORKRangeOfMotionStep.h"
+
```
Comment (umerkhan-apple) [inline ResearchKit/ActiveTasks/ORKShoulderRangeOfMotionStep.h]: new lines

File: ResearchKit/ActiveTasks/ORKShoulderRangeOfMotionStep.m [hunk 13]
```diff
@@ -0,0 +1,20 @@
+//
+//  ORKShoulderRangeOfMotionStep.m
+//  ResearchKit
+//
+//  Created by Darren Levy on 8/20/16.
+//  Copyright © 2016 researchkit.org. All rights reserved.
+//
```
Comment (umerkhan-apple) [inline ResearchKit/ActiveTasks/ORKShoulderRangeOfMotionStep.m]: legal header.

File: ResearchKit/ActiveTasks/ORKShoulderRangeOfMotionStep.m [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+
+#import "ORKShoulderRangeOfMotionStep.h"
+#import "ORKShoulderRangeOfMotionStepViewController.h"
+
+#import "ORKHelpers_Internal.h"
+
```
Comment (umerkhan-apple) [inline ResearchKit/ActiveTasks/ORKShoulderRangeOfMotionStep.m]: new lines

File: ResearchKit/ActiveTasks/ORKShoulderRangeOfMotionStep.m [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+//  ORKShoulderRangeOfMotionStep.m
+//  ResearchKit
+//
+//  Created by Darren Levy on 8/20/16.
+//  Copyright © 2016 researchkit.org. All rights reserved.
+//
+
+#import "ORKShoulderRangeOfMotionStep.h"
+#import "ORKShoulderRangeOfMotionStepViewController.h"
+
```
Comment (umerkhan-apple) [inline ResearchKit/ActiveTasks/ORKShoulderRangeOfMotionStep.m]: remove extra line

File: ResearchKit/ActiveTasks/ORKShoulderRangeOfMotionStepViewController.h [hunk 16]
```diff
@@ -0,0 +1,18 @@
+//
+//  ORKShoulderRangeOfMotionStepViewController.h
+//  ResearchKit
+//
+//  Created by Darren Levy on 8/28/16.
+//  Copyright © 2016 researchkit.org. All rights reserved.
+//
```
Comment (umerkhan-apple) [inline ResearchKit/ActiveTasks/ORKShoulderRangeOfMotionStepViewController.h]: legal header

File: ResearchKit/ActiveTasks/ORKShoulderRangeOfMotionStepViewController.h [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+
+#import "ORKRangeOfMotionStepViewController.h"
+
```
Comment (umerkhan-apple) [inline ResearchKit/ActiveTasks/ORKShoulderRangeOfMotionStepViewController.h]: new line

File: ResearchKit/ActiveTasks/ORKTouchAnywhereStep.h [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+ and the text "Touch anywhere to continue." The user can touch almost anywhere on the
+ screen and the step will end and the task will continue. The back button is still
+ tappable.
+ */
+ORK_CLASS_AVAILABLE
+@interface ORKTouchAnywhereStep : ORKActiveStep
+
+- (instancetype)initWithIdentifier:(NSString *)identifier instructionText:(NSString *)instructionText;
+
+@end
```
Comment (umerkhan-apple) [inline ResearchKit/ActiveTasks/ORKTouchAnywhereStep.h]: Make this step public. Will be very useful for other use cases as well!
Add documentation to the initializer.

Add the following:
`- (instancetype)init; NS_UNAVAILABLE`

File: ResearchKit/ActiveTasks/ORKTouchAnywhereStepViewController.m [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+#import "ORKSkin.h"
+
+@interface ORKTouchAnywhereView : ORKActiveStepCustomView {
+    NSLayoutConstraint *_topConstraint;
+}
+
+@property (nonatomic, strong) UIView *progressView;
+
+@end
+
```
Comment (umerkhan-apple) [inline ResearchKit/ActiveTasks/ORKTouchAnywhereStepViewController.m]: add extra new line

File: ResearchKit/Common/ORKOrderedTask.h [hunk 20]
```diff
@@ -206,6 +206,23 @@ typedef NS_OPTIONS(NSUInteger, ORKTremorActiveTaskOption) {
     ORKTremorActiveTaskOptionExcludeQueenWave = (1 << 4)
 } ORK_ENUM_AVAILABLE;
 
+/**
+ Values that identify the left or right limb to be used in an active task.
+*/
+typedef NS_OPTIONS(NSUInteger, ORKPredefinedTaskLimbOption) {
+    /// Which limb to use is undefined
+    ORKPredefinedTaskLimbOptionUnspecified = 0,
```
Comment (umerkhan-apple) [inline ResearchKit/Common/ORKOrderedTask.h]: Is this necessary, or can we make left be default?

File: ResearchKit/Common/ORKOrderedTask.h [hunk 21]
```diff
@@ -313,8 +330,16 @@ typedef NS_OPTIONS(NSUInteger, ORKTremorActiveTaskOption) {
                                           restDuration:(NSTimeInterval)restDuration
                                                options:(ORKPredefinedTaskOption)options;
 
++ (ORKOrderedTask *)kneeRangeOfMotionTaskWithIdentifier:(NSString *)identifier
+                                        limbOption:(ORKPredefinedTaskLimbOption)limbOption
+                                intendedUseDescription:(nullable NSString *)intendedUseDescription
+                                               options:(ORKPredefinedTaskOption)options;
```
Comment (umerkhan-apple) [inline ResearchKit/Common/ORKOrderedTask.h]: adjust spacing

File: ResearchKit/Common/ORKOrderedTask.m [hunk 22]
```diff
@@ -1022,6 +1029,139 @@ + (ORKOrderedTask *)walkBackAndForthTaskWithIdentifier:(NSString *)identifier
     return task;
 }
 
++ (ORKOrderedTask *)kneeRangeOfMotionTaskWithIdentifier:(NSString *)identifier
+                                       limbOption:(ORKPredefinedTaskLimbOption)limbOption
+                            intendedUseDescription:(NSString *)intendedUseDescription
+                                           options:(ORKPredefinedTaskOption)options {
```
Comment (umerkhan-apple) [inline ResearchKit/Common/ORKOrderedTask.m]: adjust spacing

File: ResearchKit/Common/ORKOrderedTask.m [hunk 23]
```diff
@@ -1022,6 +1029,139 @@ + (ORKOrderedTask *)walkBackAndForthTaskWithIdentifier:(NSString *)identifier
...
+    }
+    
+    ORKOrderedTask *task = [[ORKOrderedTask alloc] initWithIdentifier:identifier steps:steps];
+    return task;
+}
+
++ (ORKOrderedTask *)shoulderRangeOfMotionTaskWithIdentifier:(NSString *)identifier
+                                        limbOption:(ORKPredefinedTaskLimbOption)limbOption
+                            intendedUseDescription:(NSString *)intendedUseDescription
+                                           options:(ORKPredefinedTaskOption)options {
```
Comment (umerkhan-apple) [inline ResearchKit/Common/ORKOrderedTask.m]: adjust spacing

File: ResearchKit/Common/ORKResult.h [hunk 24]
```diff
@@ -263,6 +263,28 @@ ORK_CLASS_AVAILABLE
...
+
+ORK_CLASS_AVAILABLE
+@interface ORKRangeOfMotionResult : ORKResult
+
+/**
+ The degrees when bent.
+ */
+@property (nonatomic, assign) double flexed;
+
+
```
Comment (umerkhan-apple) [inline ResearchKit/Common/ORKResult.h]: remove extra line

File: ResearchKit/ResearchKit_Private.h [hunk 25]
```diff
@@ -48,6 +48,7 @@
 #import "ORKAudioStep.h"
 #import "ORKCompletionStep.h"
 #import "ORKCountdownStep.h"
+#import "ORKTouchAnywhereStep.h"
```
Comment (umerkhan-apple) [inline ResearchKit/ResearchKit_Private.h]: Move to public.

**Comment by umerkhan-apple** [comment]
@darrenlevy Thank you for this great contribution! Overall the tasks look amazing! Some minor formatting issues that I have pointed out.

One thing was the last step on both tasks. The checkmark is not consistent with the checkmarks shown in other RK tasks. Below is a screenshot of how it looks. I would highly recommend using an ORKCompletionStep as the last step. This way the user doesn't have to hit `Done` in the navigation bar, but instead there is already a done button included as part of the Completion Step. It would be consistent with the other Active Tasks, and more intuitive for RK users.

Apart from these things, the PR looks great. Once these changes are in, the PR will be ready to be merged! 👍 

simulator screen shot nov 7 2016 11 44 24 am

**Comment by darrenlevy** [comment]
Hi @umerkhan-apple,
Thanks so much for the comments!
I've made all the changes you specified.

Regarding whether ``ORKPredefinedTaskLimbOptionUnspecified`` is necessary:
The ``ORKPredefinedTaskLimbOption`` follows how ``ORKPredefinedTaskHandOption`` was implemented. I believe it is good to have ``ORKPredefinedTaskLimbOptionUnspecified`` be 0, not ``ORKPredefinedTaskLimbOptionLeft``  as it allows for ``ORKPredefinedTaskLimbOptionBoth = ORKPredefinedTaskLimbOptionLeft | ORKPredefinedTaskLimbOptionRight``

Also, that screenshot you posted looks consistent with the other tasks on master to me. Can you post what you're seeing for the other RK Active Tasks' Completion Step?

**Comment by umerkhan-apple** [comment]
@darrenlevy Thanks for the quick updates!

Agree with your explanation on the enum. Let's keep it as is.

I also see what you mean about the Completion Step. Must be an autolayout bug introduced with iOS 10. Will create an issue tracking it.

PR looks good! 👍 
r=me.

**Pull request merged**